python/convert_to_dataset.py

```diff
@@ -0,0 +1,128 @@
+            apiKey=os.getenv('SYN_APIKEY'),
+            silent=True)
+    except synapseclient.core.exceptions.SynapseNoCredentialsError:
+        print("Credentials not found; please manually provide your",
+              "Synapse username and password.")
+        username = input("Synapse username: ")
+        password = getpass.getpass("Synapse password: ")
+        syn = synapseclient.login(username, password, silent=True)
+    return syn
+
+
+def create_dataset_schema(syn):
+    """Create schema for creating a Dataset entity.
+
+    Returns:
+        list of ColumnModel IDS
+    """
+    cols = syn.createColumns(
+        columns=[
+            Column(name="id", columnType="ENTITYID"),
+            Column(name="fileName", columnType="STRING", maximumSize=256),
+            Column(name="name", columnType="STRING", maximumSize=256),
+            Column(name="species", columnType="STRING", maximumSize=25),
+            Column(name="dataFormat", columnType="STRING", maximumSize=15),
+            Column(name="assay", columnType="STRING", maximumSize=500),
+            Column(name="tumorType", columnType="STRING", maximumSize=500),
+            Column(name="gender", columnType="STRING", maximumSize=11)
+        ]
+    )
+    return [col.get('id') for col in cols]
+
+
+def convert_data_files(syn, folder_id):
+    """Convert `File` entities into `DatasetItem` entities.
+
+    Assumptions:
+        Immediate children of given Folder ID are File entities.
+
+    Returns:
+        list of `DatasetItem`s
+    """
+    files = syn.getChildren(folder_id)
+    return [
+        {'entityId': f.get('id'), 'versionNumber': f.get('versionLabel')}
+        for f in files
+    ]
+
+
+def convert_folder_to_dataset(syn, dataset_name, schema, dataset_items):
+    """Convert `Folder` entity into `Dataset` entity.
+
+    TODO:
+        Replace REST call with Dataset service function once available.
+    """
+    req_body = {
+        'name': dataset_name,
+        'parentId': PARENT_ID,
+        'concreteType': "org.sagebionetworks.repo.model.table.Dataset",
+        'columnIds': schema,
+        'items': dataset_items
+    }
+    syn.restPOST("/entity", json.dumps(req_body))
+
+
+def main():
+    """Main function."""
+    syn = login()
+
+    # Keep count of conversions done for summary report.
+    folder_count = 0
+    file_count = 0
+    needs_review = []
+
+    # Convert "datasets" in portal DB project.
+    col_ids = create_dataset_schema(syn)
+    for folder in syn.getChildren(FOLDERS_ID):
+        folder_id = folder.get('id')
+        files = convert_data_files(syn, folder_id)
+        try:
+            convert_folder_to_dataset(syn,
+                                      dataset_name=folder.get('name'),
+                                      schema=col_ids,
+                                      dataset_items=files)
+            folder_count += 1
+            file_count += len(files)
+        except synapseclient.core.exceptions.SynapseHTTPError:
```

Comment:
Just out of curiosity - what are same of the errors you have gotten?

Reply:
The function is assuming that all immediate children of the Folder entity are File entities.  So the error I came across was when the Dataset was being created and not all of the DatasetItems are File entities.

That said, I think I need to move my assumption from line 66.  Thanks for bringing that to my attention :sweat_smile: